apps/functions-subscription/src/use-cases/process-subscription-request.ts

```diff
@@ -0,0 +1,44 @@
+import { pipe } from 'fp-ts/lib/function';
+import * as TE from 'fp-ts/lib/TaskEither';
+import * as RTE from 'fp-ts/lib/ReaderTaskEither';
+import { SubscriptionRequest } from '../domain/subscription-request';
+import { Capabilities } from '../domain/capabilities';
+import { Subscription, makeSubscription } from '../domain/subscription';
+import { ItemAlreadyExists } from '../domain/errors';
+import { makeSubscriptionHistory } from '../domain/subscription-history';
+
+type Env = Pick<
+  Capabilities,
+  'subscriptionWriter' | 'subscriptionHistoryWriter'
+>;
+
+const recoverItemAlreadyExists =
+  (subscription: Subscription) => (error: Error) => {
+    const { trialId, userId } = subscription;
+    if (error instanceof ItemAlreadyExists)
+      return TE.right({ trialId, userId });
+    else return TE.left(error);
+  };
+
+export const processSubscriptionRequest = ({
+  trialId,
+  userId,
+}: SubscriptionRequest) =>
+  pipe(
+    RTE.ask<Env>(),
+    RTE.apSW('subscription', makeSubscription(trialId, userId)),
+    RTE.bindW('subscriptionHistory', ({ subscription }) =>
+      makeSubscriptionHistory(subscription),
+    ),
+    RTE.flatMapTaskEither((env) =>
```

Comment:
**suggestion(non-blocking)**: i'd rename the `env` variable

I think the name `env`, for a structure constructed with previous steps, could be misleading with the environment; this structure contains the `subscription` which is not a part of the environment.

Reply:
👍 , do you have any names in mind that seem clearer to you? Otherwise, we can destructure it as `RTE.flatMapTaskEither(({ subscription, subscriptionHistory, ...env }) =>`

Reply:
Nope, I'd go with the destructuring 😄